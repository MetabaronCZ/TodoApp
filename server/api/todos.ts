import { Application } from 'express';
import { InferType, boolean, object, string } from 'yup';

import { Todo } from 'models/Todo';

import { mockedDb } from '../db';
import {
  mockApiRequest,
  mockCreatedId,
  mockParseParamArray,
  mockParseParam,
  mockParseInteger,
} from '../utils';

const createTodoDataSchema = object({
  title: string().required(),
  description: string().required(),
  folder: string().nullable().required(),
  isDone: boolean().required(),
}).required();

const editTodoDataSchema = object({
  title: string(),
  description: string(),
  folder: string().nullable(),
  isDone: boolean(),
}).required();

export const setApiTodoEndpoints = (app: Application): void => {
  // get todo list
  app.get('/todo', async (req, res) => {
    await mockApiRequest();

    // parse params
    const perPage = mockParseInteger(req.query.perPage) ?? 20;
    const page = mockParseInteger(req.query.page) ?? 0;
    const folder = mockParseParam(req.query.folder);
    const query = mockParseParam(req.query.query);
    const sort = mockParseParam(req.query.sort);

    // filter items
    const items = mockedDb.todos.filter((item) => {
      // filter by folder
      if (folder && folder !== item.folder) {
        return false;
      }
      // filter by query string
      if (
        query &&
        !item.title.toLowerCase().includes(query.toLowerCase()) &&
        !item.description.toLowerCase().includes(query.toLowerCase())
      ) {
        return false;
      }
      return true;
    });

    // sort items
    items.sort((a, b) => {
      switch (sort) {
        case 'TITLE_ASC':
          return a.title.localeCompare(b.title);
        case 'TITLE_DESC':
          return b.title.localeCompare(a.title);
        case 'CREATED_ASC':
          return a.created - b.created;
        case 'CREATED_DESC':
          return b.created - a.created;
        case 'DONE_1':
          return a.isDone === b.isDone ? 0 : a.isDone ? -1 : +1;
        case 'DONE_0':
          return a.isDone === b.isDone ? 0 : a.isDone ? +1 : -1;
        default:
          // use CREATED_DESC as default
          return b.created - a.created;
      }
    });

    const start = page * perPage;
    const end = start + perPage;

    res.json({
      data: {
        items: items.slice(start, end),
        count: items.length,
      },
    });
  });

  // get todo detail
  app.get('/todo/:id', async (req, res) => {
    const { id } = req.params;
    await mockApiRequest();

    const todo = mockedDb.todos.find((item) => id === item.id);
    res.json({ data: todo || null });
  });

  // create todo
  app.post('/todo', async (req, res) => {
    const data = req.body;
    await mockApiRequest();

    try {
      await createTodoDataSchema.validate(data);
      const todoData = data as InferType<typeof createTodoDataSchema>;

      const todo: Todo = {
        ...todoData,
        id: mockCreatedId('TODO'),
        created: Date.now(),
      };
      mockedDb.todos = [...mockedDb.todos, todo];
      res.json({ data: todo });
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // edit todo
  app.patch('/todo/:id', async (req, res) => {
    const { id } = req.params;
    const data = req.body;
    await mockApiRequest();

    try {
      await editTodoDataSchema.validate(data);
      const todoData = data as InferType<typeof editTodoDataSchema>;

      mockedDb.todos = mockedDb.todos.map((item) => {
        return id === item.id ? { ...item, ...todoData } : item;
      });

      res.json(null);
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // delete todos
  app.delete('/todo', async (req, res) => {
    const { ids } = req.query;
    await mockApiRequest();

    const idsParsed = mockParseParamArray(ids);

    const validIds = mockedDb.todos
      .filter((item) => idsParsed.includes(item.id))
      .map((item) => item.id);

    if (0 === validIds.length) {
      res.status(500).json({ error: 'Could not delete todos: Invalid ID!' });
    } else {
      mockedDb.todos = mockedDb.todos.filter((item) => {
        return !validIds.includes(item.id);
      });

      res.json(null);
    }
  });
};
