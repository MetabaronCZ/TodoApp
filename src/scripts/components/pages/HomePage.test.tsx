import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import { HomePage } from 'components/pages/HomePage';

import { mockStore } from 'test/store';
import { withMockedProviders } from 'test/component';
import * as actionsModule from 'store/todos/actions';

const getHomepage = (searchParams?: Record<string, string>): JSX.Element => {
  const store = mockStore();
  const search = new URLSearchParams(searchParams);

  return withMockedProviders(
    <Provider store={store}>
      <HomePage />
    </Provider>,
    {
      url: search ? `?${search.toString()}` : null,
    },
  );
};

describe('components/pages/HomePage', () => {
  it('should render correctly', async () => {
    const { container } = render(getHomepage());
    expect(container).toMatchSnapshot();
  });

  it('should call refetch when no folder param given', async () => {
    const mockedFetch = jest.spyOn(actionsModule, 'fetchTodos');
    const mockedFilter = jest.spyOn(actionsModule, 'filterTodos');
    expect(mockedFetch.mock.calls.length).toEqual(0);
    expect(mockedFilter.mock.calls.length).toEqual(0);

    render(getHomepage());
    expect(mockedFetch.mock.calls.length).toEqual(1);
    expect(mockedFilter.mock.calls.length).toEqual(0);
  });

  it('should filter items when folder param given', async () => {
    const mockedFetch = jest.spyOn(actionsModule, 'fetchTodos');
    const mockedFilter = jest.spyOn(actionsModule, 'filterTodos');
    expect(mockedFetch.mock.calls.length).toEqual(0);
    expect(mockedFilter.mock.calls.length).toEqual(0);

    render(getHomepage({ folder: 'TEST' }));
    expect(mockedFetch.mock.calls.length).toEqual(0);
    expect(mockedFilter.mock.calls.length).toEqual(1);
    expect(mockedFilter.mock.calls[0][0].folder).toEqual('TEST');
  });
});
