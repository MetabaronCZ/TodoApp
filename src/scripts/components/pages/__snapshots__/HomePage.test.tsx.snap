// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/HomePage should render correctly 1`] = `
.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 10px 0;
}

.c7 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c29 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c30 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c30:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c30:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c30:disabled {
  color: #c0c0c0;
}

.c6 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c6:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c6:disabled {
  color: #c0c0c0;
}

.c13 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c13:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c13:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c13:disabled {
  color: #c0c0c0;
}

.c15 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  padding: 0 10px;
  font-size: 20px;
  line-height: 40px;
  font-weight: bold;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 0;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
}

.c34 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #22c;
}

.c34:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c34:focus {
  outline: 1px dotted #22c;
  outline-offset: -1px;
}

.c33 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  padding: 20px 0;
  text-align: center;
}

.c0 {
  min-height: 100vh;
  width: 100%;
  padding: 0 20px;
  background: #f0f0f0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 320px;
  max-width: 980px;
  min-height: 100vh;
  width: 100%;
  margin: 0 auto;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 20px;
}

.c11 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c12 {
  border-bottom: 1px solid #c0c0c0;
}

.c12:last-child {
  border-bottom: none;
}

.c14 {
  width: 100%;
  background: #c0c0c0;
  text-align: left;
}

.c14:hover {
  background-color: #c0c0c0;
}

.c16 {
  width: 100%;
  text-align: left;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 20px;
}

.c10 {
  width: 240px;
}

.c18 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 0;
}

.c20 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 18px;
  line-height: 30px;
  font-weight: normal;
  margin: 0;
}

.c31 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

.c26 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #c0c0c0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: 30px;
}

.c24 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  width: 20px;
  height: 20px;
  border: 2px solid #c0c0c0;
  color: #c0c0c0;
  background: #fff;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c25 {
  position: absolute;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
  opacity: 0;
  outline: none;
}

.c22 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: calc(30px + 1px);
  border-bottom: 1px solid #c0c0c0;
  background: #f0f0f0;
}

.c27 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c28 {
  position: relative;
  display: inline-block;
}

.c19 {
  opacity: 0.5;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
}

.c21 {
  margin-bottom: 1px;
}

.c32 {
  padding-top: 10px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <header
        class="c2"
      >
        <a
          class="c3"
          href="/"
        >
          ☑ 
          TODOLIST
        </a>
        <div
          class="c4"
        />
        <div
          class="c5"
        >
          <a
            class="c6"
            href="/folder/list"
            title="Složky"
          >
            <i
              class="c7"
            >
              ★
            </i>
          </a>
          <a
            class="c6"
            href="/settings"
            title="Nastavení"
          >
            <i
              class="c7"
            >
              ✎
            </i>
          </a>
        </div>
      </header>
      <main
        class="c8"
      >
        <div
          class="c9"
        >
          <nav
            class="c10"
          >
            <ul
              class="c11"
            >
              <li
                class="c12"
              >
                <a
                  class="c13 c14"
                  href="/?folder="
                >
                  <i
                    class="c7"
                  >
                    ★
                  </i>
                  <span
                    class="c15"
                  >
                    Všechny úkoly
                  </span>
                </a>
              </li>
              <li
                class="c12"
              >
                <a
                  class="c13 c16"
                  href="/folder/list"
                >
                  <i
                    class="c7"
                  >
                    ✎
                  </i>
                  <span
                    class="c15"
                  >
                    Složky
                  </span>
                </a>
              </li>
            </ul>
          </nav>
          <div
            class="c17 c18"
          >
            <div
              class="c19"
            >
              <h1
                class="c20 c21"
              >
                Úkoly
              </h1>
              <div
                class="c22"
              >
                <div>
                  <div
                    class="c23"
                  >
                    <label
                      class="c24"
                      for="checkbox-2"
                    >
                      <input
                        class="c25"
                        disabled=""
                        id="checkbox-2"
                        type="checkbox"
                      />
                    </label>
                    <label
                      class="c26"
                      for="checkbox-2"
                    >
                      Vybrat vše
                    </label>
                  </div>
                </div>
                <div
                  class="c27"
                />
                <div>
                  <a
                    class="c13"
                    href="/todo/create"
                  >
                    <i
                      class="c7"
                    >
                      +
                    </i>
                    <span
                      class="c15"
                    >
                      Vytvořit
                    </span>
                  </a>
                </div>
                <div>
                  <div
                    class="c28"
                  >
                    <button
                      class="c29 c30"
                      disabled=""
                      type="button"
                    >
                      <span
                        class="c15"
                      >
                        Nejnovější
                      </span>
                      <i
                        class="c7"
                        style="color: rgb(192, 192, 192);"
                      >
                        ˅
                      </i>
                    </button>
                  </div>
                </div>
              </div>
              <p
                class="c31 c32"
              >
                Nebyly nalezeny žádné úkoly.
              </p>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="c33"
      >
        <div
          class=""
        >
          Copyright © 2023 MetabaronCZ (
          <a
            class="c34"
            href="https://github.com/MetabaronCZ"
            target="_blank"
          >
            https://github.com/MetabaronCZ
          </a>
          ). All Rights Reserved.
        </div>
      </footer>
    </div>
  </div>
</div>
`;
